operators/src/processing/raster_kernel/kernel_function.rs

```diff
@@ -0,0 +1,208 @@
+use super::{error, RasterKernelError};
+use crate::util::{number_statistics::NumberStatistics, Result};
+use geoengine_datatypes::raster::{Grid2D, GridShape2D, GridSize, Pixel};
+use num::Integer;
+use num_traits::AsPrimitive;
+use snafu::ensure;
+
+/// A function that calculates one pixel values from a neighborhood of pixels.
+pub trait KernelFunction<P>: Sync + Send + Clone
+where
+    P: Pixel,
+    f64: AsPrimitive<P>,
+{
+    fn kernel_size(&self) -> GridShape2D;
+
+    /// Specifies the x extent beginning from the center pixel
+    fn x_radius(&self) -> usize {
+        self.kernel_size().axis_size_x() / 2
+    }
+
+    fn x_width(&self) -> usize {
+        self.kernel_size().axis_size_x()
+    }
+
+    /// Specifies the y extent beginning from the center pixel
+    fn y_width(&self) -> usize {
+        self.kernel_size().axis_size_y()
+    }
+
+    /// Specifies the x extent right of one pixel
+    fn y_radius(&self) -> usize {
+        self.kernel_size().axis_size_y() / 2
+    }
+
+    fn apply(&self, values: &[Option<P>]) -> Option<P>;
```

Comment:
add documentation? isn't there a way to enforce the values to have the correct number of elements? What happens if it's not the correct one? panic?

Reply:
changed this in [b6b3237](https://github.com/geo-engine/geoengine/pull/644/commits/b6b323705d0d990d2ce33577c0f17192939cefcb)